reader_concurrency_semaphore.cc

```diff
@@ -1183,7 +1181,20 @@ future<> reader_concurrency_semaphore::request_memory(reader_permit::impl& permi
         return make_ready_future<>();
     }
 
-    return enqueue_waiter(reader_permit(permit.shared_from_this()), wait_on::memory);
+    return enqueue_waiter(permit, wait_on::memory);
+}
+
+void reader_concurrency_semaphore::dequeue_permit(reader_permit::impl& permit) {
+    switch (permit.get_state()) {
+        case reader_permit::state::waiting_for_admission:
+        case reader_permit::state::waiting_for_memory:
+            --_stats.waiters;
+            break;
+        default:
+            on_internal_error_noexcept(rcslog, format("reader_concurrency_semaphore::dequeue_permit(): unrecognized queued state: {}", permit.get_state()));
+    }
```

Comment:
Better not to have a default, so the compiler can complain.

Reply:
Done in v2.